pyquil/api/_client.py

```diff
@@ -0,0 +1,241 @@
+            self._config = QCSClientConfiguration.load()
+
+    @property
+    def qvm_url(self) -> str:
+        """
+        QVM URL from client configuration.
+        """
+        return self._config.profile.applications.pyquil.qvm_url
+
+    @property
+    def quilc_url(self) -> str:
+        """
+        Quil compiler URL from client configuration.
+        """
+        return self._config.profile.applications.pyquil.quilc_url
+
+    def qvm_version(self) -> str:
+        """
+        Get QVM version string.
+        """
+        response = self.post_json(self.qvm_url, {"type": "version"})
+        split_version_string = response.text.split()
+        try:
+            qvm_version = split_version_string[0]
+        except ValueError:
+            raise TypeError(f"Malformed version string returned by the QVM: {response.text}")
+        return qvm_version
+
+    @contextmanager
+    def _http_client(self) -> Iterator[httpx.Client]:
+        if self._http is None:
+            with build_sync_client(configuration=self._config) as client:  # type: httpx.Client
+                yield client
+        else:
+            yield self._http
+
+    def _create_engagement(self, processor_id: str) -> EngagementWithCredentials:
+        return cast(
+            EngagementWithCredentials,
+            self.qcs_request(
+                create_engagement,
+                json_body=CreateEngagementRequest(quantum_processor_id=processor_id),
+            ),
+        )
+
+
+def _engagement_valid(processor_id: str, engagement: Optional[EngagementWithCredentials]) -> bool:
+    if engagement is None:
+        return False
+
+    return all(
+        [
+            engagement.credentials.client_public != "",
+            engagement.credentials.client_secret != "",
+            engagement.credentials.server_public != "",
+            parsedate(engagement.expires_at) > datetime.now(tzutc()),
+            engagement.address != "",
+            engagement.quantum_processor_id == processor_id,
+        ]
+    )
+
+
+def _to_auth_config(credentials: EngagementCredentials) -> ClientAuthConfig:
+    return rpcq.ClientAuthConfig(
+        client_secret_key=credentials.client_secret.encode(),
+        client_public_key=credentials.client_public.encode(),
+        server_public_key=credentials.server_public.encode(),
+    )
+
+
+def _parse_error(res: httpx.Response) -> ApiError:
+    """
+    Errors should contain a "status" field with a human readable explanation of
+    what went wrong as well as a "error_type" field indicating the kind of error that can be mapped
+    to a Python type.
+
+    There's a fallback error UnknownApiError for other types of exceptions (network issues, api
+    gateway problems, etc.)
+    """
+    try:
+        body = res.json()
+    except JSONDecodeError:
+        raise UnknownApiError(res.text)
+
+    if "error_type" not in body:
+        raise UnknownApiError(str(body))
+
+    error_type = body["error_type"]
```

Comment:
This function is only used when calling `Client.post_json()`, which in turn is only used against the QVM. Since I don't expect that API to have changed, I think we need to keep this error parsing.

For QCS requests, we're always using the built-in generated request functions (e.g. `create_engagement`), which do their own error handling.